internal/object/object.go

```diff
@@ -17,39 +19,60 @@ import (
 	"sync"
 )
 
+var (
+	cache   = make(map[string]*Object)
+	cacheMu sync.Mutex
+)
+
 type Object struct {
 	ID       types.Binary
+	SourceId int64
+	Name     string
 	Tags     map[string]string
-	Metadata map[int64]*SourceMetadata
+	URL      string
+
+	ExtraTags map[string]string
 
 	db *icingadb.DB
 
 	mu sync.Mutex
 }
 
-var (
-	cache   = make(map[string]*Object)
-	cacheMu sync.Mutex
-)
+func NewObject(db *icingadb.DB, ev *event.Event) *Object {
+	return &Object{
+		SourceId:  ev.SourceId,
+		Name:      ev.Name,
+		db:        db,
+		URL:       ev.URL,
+		Tags:      ev.Tags,
+		ExtraTags: ev.ExtraTags,
+	}
+}
 
-func FromTags(ctx context.Context, db *icingadb.DB, tags map[string]string) (*Object, error) {
-	id := ID(tags)
+func FromTags(ctx context.Context, db *icingadb.DB, ev *event.Event) (*Object, error) {
+	id := ID(ev.SourceId, ev.Tags)
+	object := NewObject(db, ev)
+	object.ID = id
 
 	cacheMu.Lock()
 	defer cacheMu.Unlock()
 
-	object, ok := cache[id.String()]
+	objectCache, ok := cache[id.String()]
 	if ok {
-		return object, nil
+		objectCache.ExtraTags = object.ExtraTags
+
+		return objectCache, nil
 	}
 
-	object = &Object{ID: id, Tags: tags, db: db}
 	cache[id.String()] = object
```

Comment:
Please try to keep the structure so that the cache lookup happens first and depending on whether it's in there, use it or create it. I don't see why you'd want to do `NewObject()` and then dismiss it later (and having that just makes it easier to introduce bugs where the wrong one is returned or something like that).